stream-chat-android-offline/src/main/java/io/getstream/chat/android/livedata/controller/ChannelControllerImpl.kt

```diff
@@ -608,26 +614,51 @@ internal class ChannelControllerImpl(
 
         return if (online) {
             // upload attachments
-            logger.logI("Uploading attachments for message with id ${newMessage.id} and text ${newMessage.text}")
-            newMessage.attachments = newMessage.attachments.map {
-                var attachment: Attachment = it
-                if (it.upload != null) {
-                    val result = uploadAttachment(it, attachmentTransformer)
-                    attachment.uploadComplete = result.isSuccess
-                    if (result.isSuccess) {
-                        attachment = result.data()
-                    } else {
-                        attachment.uploadState = Attachment.UploadState.Failed(result.error())
-                        logger.logE("Failed to upload attachment for message")
+
+            val hasAttachments = newMessage.attachments.isNotEmpty()
+
+            if (hasAttachments) {
+                logger.logI("Uploading attachments for message with id ${newMessage.id} and text ${newMessage.text}")
+
+                val uploadId = generateUploadId()
+                newMessage.uploadId = uploadId
+
+                val progressTracker = ProgressTrackerFactory.getOrCreate(uploadId)
+                    .apply { initProgress(newMessage.attachments.size) }
+
+                upsertProgressMessage(uploadId)
+                newMessage.attachments = newMessage.attachments.map {
+                    progressTracker.setProgress(0)
+                    var attachment: Attachment = it
+
+                    if (it.upload != null) {
+                        val result = uploadAttachment(
+                            it,
+                            attachmentTransformer,
+                            progressTracker.toProgressCallback()
+                        )
+                        attachment.uploadComplete = result.isSuccess
+
+                        if (result.isSuccess) {
+                            attachment = result.data()
+                            progressTracker.incrementLap()
+                        } else {
+                            attachment.uploadState = Attachment.UploadState.Failed(result.error())
+                        }
                     }
-                }
-                attachment
-            }.toMutableList()
 
-            logger.logI("Starting to send message with id ${newMessage.id} and text ${newMessage.text}")
+                    attachment
+                }.toMutableList()
+
+                if (hasAttachments) {
+                    cancelProgressMessage()
```

Comment:
Looking at the usage here, this is more `clear` than `cancel` (as in, it's not an exceptional case, because we completed the upload successfully). Also, it's a single-line method, maybe we can just inline it here, as we don't use it anywhere else.